server/app/api/v3.py

```diff
@@ -323,6 +389,38 @@ def register_api_v3(app, app_config, api_url_prefix):
 
         # Custom CXGs (only for "w" dataroot)
         if url_dataroot == "w":
+            # Add authorization check before processing any /w/ requests
```

Comment:
Adds a guard for /w dataroot access: requires an authenticated user and (outside local dev) enforces that userId appears in the path (or decoded path) to prevent cross-user access.